assets/js/base/hooks/checkout/use-checkout-address.js

```diff
@@ -46,55 +43,54 @@ export const useCheckoutAddress = () => {
 			( ! customerId || isSameAddress( shippingAddress, billingData ) )
 	);
 
-	// Pushes to global state when changes are made locally.
-	useEffect( () => {
-		// Uses shipping address or billing fields depending on shippingAsBilling checkbox, but ensures
-		// billing only fields are also included.
-		const newBillingData = {
-			...( shippingAsBilling ? shippingAddress : billingFields ),
-			email: billingFields.email || billingData.email,
-			phone: billingFields.phone || billingData.phone,
-		};
-
-		if ( ! isEqual( newBillingData, billingData ) ) {
-			setBillingData( newBillingData );
-		}
-	}, [
-		billingFields,
-		shippingAsBilling,
-		billingData,
-		shippingAddress,
-		setBillingData,
-		setShippingAddress,
-	] );
+	const currentShippingAsBilling = useRef( shippingAsBilling );
+	const previousBillingData = useRef( billingData );
 
 	/**
-	 * Wrapper for updateBillingFields (from useState) which handles merging.
-	 *
-	 * @param {Object} newValues New values to store to state.
+	 * Sets shipping address data, and also billing if using the same address.
 	 */
-	const setBillingFields = useCallback(
-		( newValues ) =>
-			void updateBillingFields( ( prevState ) => ( {
-				...prevState,
-				...newValues,
-			} ) ),
-		[]
+	const setShippingFields = useCallback(
+		( value ) => {
+			setShippingAddress( value );
+
+			if ( shippingAsBilling ) {
+				setBillingData( value );
+			}
+		},
+		[ shippingAsBilling, setShippingAddress, setBillingData ]
 	);
 
-	const setEmail = ( value ) => void setBillingFields( { email: value } );
-	const setPhone = ( value ) => void setBillingFields( { phone: value } );
+	// This syncs billing data with shipping data if the checkbox is enabled.
+	useEffect( () => {
+		if ( currentShippingAsBilling.current !== shippingAsBilling ) {
```

Comment:
So just for clarity sake - I was originally wondering why `shippingAsBilling` was being tracked as a ref but it looks like why it's implemented this way is because:

- the `billingData` dependency could trigger this effect frequently (because it's an object that changes).
- tracking previous state of `shippingAsBilling` allows for escaping the execution of this logic and only doing so when the `shippingAsBilling` state changes.

The potential problem here is that we still want billing data to update when shipping data changes when `shippingAsBilling` is true but this problem is resolved  by the `setShippingFields` callback.

Is the above correct?

Reply:
Right. This effect basically only handles the toggle event; when the checkbox is toggled we need to immediately update billing address to reflect this change. I will add a comment.